File: src/#16/solution.go
package solution

import (
	"math"
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	n, res, diff := len(nums), 0, math.MaxInt32

	if n > 2 {
		sort.Ints(nums)
		for i := 0; i < n-2; i++ {
			for left, right := i+1, n-1; left < right; {
				sum := nums[i] + nums[left] + nums[right]
				if abs(sum-target) < diff {
					res, diff = sum, abs(sum-target)
				}
				if sum == target {
					return res
				} else if sum > target {
					right--
				} else {
					left++
				}
			}
		}
	}
	return res
}

func abs(num int) int {
	if num < 0 {
		return num * -1
	}
	return num
}

File: src/#16/solution.js
var threeSumClosest = function (nums, target) {
  let n = nums.length
  let res = 0
  let diff = Number.MAX_VALUE
  if (n > 2) {
    nums.sort((a, b) => a - b)

    for (let i = 0; i < n - 2; i++) {
      let left = i + 1,
        right = n - 1

      while (left < right) {
        let sum = nums[i] + nums[left] + nums[right]

        if (Math.abs(sum - target) < diff) {
          diff = Math.abs(sum - target)
          res = sum
        }

        if (sum === target) {
          return res
        } else if (sum > target) {
          right--
        } else {
          left++
        }
      }
    }
  }
  return res
}
